### PR TITLE
Addition of Nickname Moderation CC

### PR DESCRIPTION
**Pull request by 2x2master1**
**Please describe the changes this PR makes and why it should be merged:**
I made a custom command on nickname moderation, more details on the CC itself can be seen in the comment at the beginning.
I would appreciate if it would be put into the repository.

**Status**

- [x] Code changes have been tested on an instance of YAGPDB, or there are no code changes
- [x] I have read and followed the [contribution guide](../CONTRIBUTING.md)

### CODE REVIEW

**Comment by 2x2master1** [comment]
By the way, this is the first pull request I have ever done, so apologies if something is wrong.

**Comment by DaviiD1337** [comment]
Hmm, you are never checking for arg no. 1 (index .CmdArgs 1) so it errors >
image. Also I prefer so the prefix is not a manual variable.

File: moderation/nickname-moderation.go.tmpl [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	-modnick <@user/ID>
+	-unmodnick <@user/ID>
+	-nick-numberreset
+
+	Credit: 2x2Master#7363
+	Feel free to ping me in YAGPDB support if you have any questions or problems!
+*/}}
+
+{{/* Configuration area starts */}}
+{{$prefix := "-"}}{{/* YAGPDB's prefix in your server */}}
```
Comment (DaviiD1337) [inline moderation/nickname-moderation.go.tmpl]: Use this instead > 
```{{$prefix := index (reFindAllSubmatches `Prefix of \x60\d+\x60: \x60(.+)\x60` (exec "prefix")) 0 1}}```

File: moderation/nickname-moderation.go.tmpl [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+{{$mod := false}}
+{{range $modRoles -}}
+	{{if hasRoleID .}}
+		{{$mod = true}}
+	{{end}}
+{{end}}
+{{if eq (lower (index .CmdArgs 0)) (print $prefix "modnick")}}
+	{{if $mod}}
+	{{((userArg (index .CmdArgs 1)).String)}}'s nickname will be moderated when they next send a message.
```
Comment (DaviiD1337) [inline moderation/nickname-moderation.go.tmpl]: You are never checking for an existence of (index .CmdArgs 1). You can do this like so:
```{{if ge (len .CmdArgs) 2}}```. Same for all commands.

File: moderation/nickname-moderation.go.tmpl [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	{{end}}
+{{else if (eq (lower (index .CmdArgs 0)) (print $prefix "nick-numberreset"))}}
+	{{if $mod}}
+	{{dbSet 0 "mnick" 0}}
+	{{addReactions "✅"}}
+	{{else}}
+	{{sendMessage nil (print .User.Mention ", you can't use that.")}}
+	{{end}}
+{{else if eq (toInt ((dbGet .User.ID "nick").Value)) 1}}
+	{{if or (ne (reFind $name .Member.Nick) $name) (eq .Member.Nick "")}}
```
Comment (DaviiD1337) [inline moderation/nickname-moderation.go.tmpl]: I'd prefer using >
```(not .Member.Nick)```
also all line:
```{{if or (reFind $name .Member.Nick) (not .Member.Nick) }}```

File: moderation/nickname-moderation.go.tmpl [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	{{$db := dbIncr 0 "mnick" 1}}
+	{{$db2 := toInt (print (dbGet 0 "mnick").Value)}}
```
Comment (DaviiD1337) [inline moderation/nickname-moderation.go.tmpl]: ```{{ toInt $db }}``` 😳

File: moderation/NicknameModeration.go.tmpl [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	-modnick <@user/ID>
+	-unmodnick <@user/ID>
+	-nick-numberreset
+
+	Credit: 2x2Master#7363
+	Feel free to ping me in YAGPDB support if you have any questions or problems!
+*/}}
+
+{{/* Configuration area starts */}}
+{{$prefix := "-"}}{{/* YAGPDB's prefix in your server */}}
```
Comment (l-zeuch) [inline moderation/NicknameModeration.go.tmpl:21]: Would remove this line and add it further down, filling the variable with the following:
```go
{{$prefix := index (reFindAllSubmatches `.*?: \x60(.*)\x60\z` (execAdmin "prefix")) 0 1}}
```` 
As David said, I'm not a fan of the prefix as a user variable either...

File: moderation/NicknameModeration.go.tmpl [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+	Trigger type: Regex
+	Trigger: \A
+
+	Usage:
+	-modnick <@user/ID>
+	-unmodnick <@user/ID>
+	-nick-numberreset
+
+	Credit: 2x2Master#7363
```
Comment (l-zeuch) [inline moderation/NicknameModeration.go.tmpl:16]: If you really want, we could keep it like that, but it is recommended to instead use your GitHub Account, as keeping up with potential tag or username changes on Discord is rather tedious:
```suggestion
	Credits: 2x2Master1 <https://github.com/2x2master1>
```

File: moderation/NicknameModeration.go.tmpl [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+{{$modRoles := cslice 809781706658807838 815197447000555530}}{{/* Users with these roles can moderate nicknames */}}
+{{/* Configuration area ends */}}
+
+
+{{$mod := false}}
+{{range $modRoles -}}
+	{{if hasRoleID .}}
+		{{$mod = true}}
+	{{end}}
+{{end}}
```
Comment (l-zeuch) [inline moderation/NicknameModeration.go.tmpl:33]: Why not go with a permission check? Would imo be a more fitting and easier for the users:
```suggestion
{{$mod := in (split (index (split (exec "viewperms") "\n") 2) ", ") "ManageMessages"}}
```

**Review by DaviiD1337 — CHANGES_REQUESTED** [review]

**Review by l-zeuch — CHANGES_REQUESTED** [review]
Good first PR, just a few things. (Don't worry, it's not much :D)

**Comment by 2x2master1** [comment]
Thank you for the super helpful tips! I made some edits to the code. Am I all set now?

**Comment by l-zeuch** [comment]
I'll take a quick look, just a moment.
Thanks for the fast reply!

**Comment by 2x2master1** [comment]
There will be an error when sending an image without message text, this can be prevented by disabling Errors as CC response however.

**Comment by l-zeuch** [comment]
And error handling is not possible? Sounds like a hack to me...

**Comment by 2x2master1** [comment]
It is, I'll add it real quick. `{{if .CmdArgs}}` should do the job.

**Comment by l-zeuch** [comment]
Gotcha, then add it and I'll review your PR 👍🏾

**Comment by 2x2master1** [comment]
Done.

**Comment by l-zeuch** [comment]
Very nice, let me take a look.

**Comment by 2x2master1** [comment]
Looks like the fix is not working, it errors when saving. Sorry, I'll have a look again.

**Comment by l-zeuch** [comment]
Hit me up when you've fixed it! I'll wait for the time being.

**Comment by 2x2master1** [comment]
I've done some testing, it's not giving errors anymore. This should be it 👍

**Review by l-zeuch — APPROVED** [review]
lgtm, if you're ready to merge, mark this PR as ready!

**Comment by l-zeuch** [comment]
@DaviiD1337 , do you want to go over it or is it ready to merge?

**Pull request merged**